systems/controllers/osc/osc_tracking_data_mbp.h

```diff
@@ -0,0 +1,377 @@
+
+  // Update() updates the caches. It does the following things in order:
+  //  - update track_at_current_state_
+  //  - update desired output
+  //  - update feedback output (Calling virtual methods)
+  //  - update command output (desired output with pd control)
+  // Inputs/Arguments:
+  //  - `x_w_spr`, state of the robot (with spring)
+  //  - `cache_w_spr`, plant context of the robot (without spring)
+  //  - `x_wo_spr`, state of the robot (with spring)
+  //  - `context_wo_spr`, plant context of the robot (without spring)
+  //  - `traj`, desired trajectory
+  //  - `t`, current time
+  //  - `finite_state_machine_state`, current finite state machine state
+  bool Update(const Eigen::VectorXd& x_w_spr,
+              drake::systems::Context<double>& context_w_spr,
+              const Eigen::VectorXd& x_wo_spr,
+              drake::systems::Context<double>& context_wo_spr,
+              const drake::trajectories::Trajectory<double>& traj, double t,
+              int finite_state_machine_state);
+  // Getters used by osc block
+  Eigen::VectorXd GetOutput() { return y_; }
+  Eigen::MatrixXd GetJ() { return J_; }
+  Eigen::VectorXd GetJdotTimesV() { return JdotV_; }
+  Eigen::VectorXd GetDdyCommand() { return ddy_command_; }
+  Eigen::MatrixXd GetWeight() { return W_; }
+  // void UpdatePGain(Eigen::MatrixXd K_p) {K_p_ = K_p;}
+  // void UpdateDGain(Eigen::MatrixXd K_d) {K_d_ = K_d;}
+  // void UpdateWeight(Eigen::MatrixXd W);
+
+  // Getters
+  std::string GetName() { return name_; };
+  int GetTrajDim() { return n_r_; };
+  bool IsActive() { return track_at_current_state_; }
+
+  void SaveDdyCommandSol(const Eigen::VectorXd& dv);
+
+  // Print feedback and desired values
+  void PrintFeedbackAndDesiredValues(const Eigen::VectorXd& dv);
+
+  // Finalize and ensure that users construct OscTrackingDataMBP class
+  // correctly.
+  void CheckOscTrackingData();
+
+ protected:
+  int GetStateIdx() { return state_idx_; };
+  void AddState(int state);
+
+  // Feedback output, Jacobian and dJ/dt * v
+  Eigen::VectorXd error_y_;
+  Eigen::VectorXd error_dy_;
+  Eigen::VectorXd y_;
+  Eigen::VectorXd dy_;
+  Eigen::MatrixXd J_;
+  Eigen::VectorXd JdotV_;
+
+  // Desired output
+  Eigen::VectorXd y_des_;
+  Eigen::VectorXd dy_des_;
+  Eigen::VectorXd ddy_des_;
+  Eigen::VectorXd ddy_des_converted_;
+
+  // Commanded acceleration after feedback terms
+  Eigen::VectorXd ddy_command_;
+  // Osc solution
+  Eigen::VectorXd ddy_command_sol_;
+
+  // `state_` is the finite state machine state when the tracking is enabled
+  // If `state_` is empty, then the tracking is always on.
+  std::vector<int> state_;
+
+  /// OSC calculates feedback positions/velocities from `tree_w_spr_`,
+  /// but in the optimization it uses `tree_wo_spr_`. The reason of using
+  /// RigidBodyTree without spring is that the OSC cannot track desired
+  /// acceleration instantaneously when springs exist. (relative degrees of 4)
```

Comment:
Should update the documentation (both Osc and OscTrackingData)

Reply:
Done.